bitsail-connectors/connector-localfilesystem/src/main/java/com/bytedance/bitsail/connector/localfilesystem/reader/LocalFileSystemSourceReader.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2022 Bytedance Ltd. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.bytedance.bitsail.connector.localfilesystem.reader;
+
+import com.bytedance.bitsail.base.connector.reader.v1.SourcePipeline;
+import com.bytedance.bitsail.base.format.DeserializationSchema;
+import com.bytedance.bitsail.common.configuration.BitSailConfiguration;
+import com.bytedance.bitsail.common.row.Row;
+import com.bytedance.bitsail.connector.base.source.SimpleSourceReaderBase;
+import com.bytedance.bitsail.connector.localfilesystem.core.config.FileSystemConfig;
+
+import java.io.BufferedReader;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class LocalFileSystemSourceReader extends SimpleSourceReaderBase<Row> {
+  private final FileSystemConfig fileSystemConfig;
+  private final String filePath;
+  private final BufferedReader bufferedReader;
+  private final transient DeserializationSchema<byte[], Row> deserializationSchema;
+
+  public LocalFileSystemSourceReader(BitSailConfiguration jobConf, Context context) {
+    this.fileSystemConfig = new FileSystemConfig(jobConf);
+    this.filePath = fileSystemConfig.getFilePath();
+    this.bufferedReader = loadCsvFile();
+    this.deserializationSchema = DeserializationSchemaFactory.createDeserializationSchema(
+        jobConf,
+        context,
+        fileSystemConfig
+    );
+  }
+
+  private BufferedReader loadCsvFile() {
+    Path path = Paths.get(this.filePath);
+    if (!Files.exists(path)) {
+      throw new RuntimeException(new FileNotFoundException(
+        String.format("File %s does not exits!", this.filePath)
+      ));
+    }
+
+    BufferedReader bufferedReader;
+    try {
+      bufferedReader = Files.newBufferedReader(path);
+      if (this.fileSystemConfig.getSkipFirstLine()) {
+        bufferedReader.readLine();
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    return bufferedReader;
+  }
+
+  @Override
+  public void pollNext(SourcePipeline<Row> pipeline) throws Exception {
+    String line = this.bufferedReader.readLine();
+    Row row = deserializationSchema.deserialize(line.getBytes());
+    pipeline.output(row);
+  }
+
+  @Override
+  public boolean hasMoreElements() {
+    try {
+      boolean isReady = this.bufferedReader.ready();

```

Comment:
what does `ready` mean here? it means it reach the end of file?

Reply:
Method `ready()` can tell us whether the buffer stream is ready to read. After looking through tech blog, I realize that it may be incorrect to determine whether it reaches the end of file. So I will choose another way to implement `hasMoreElements()`.